src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -270,26 +278,50 @@ data Level m h = Level {
 -- | A merging run is either a single run, or some ongoing merge.
 data MergingRun m h =
     -- | A merging of multiple runs.
-    MergingRun !MergePolicyForLevel !NumRuns !(StrictMVar m (MergingRunState m h))
+    MergingRun
+      !MergePolicyForLevel
+      !NumRuns
+      !NumEntries
+      -- ^ Sum of number of entries in the input runs
+      !(UnspentCreditsVar (PrimState m))
+      -- ^ The number of currently /unspent/ credits
+      !(TotalStepsVar (PrimState m))
+      -- ^ The total number of performed merging steps.
+      !(MutVar (PrimState m) MergeCompleted)
+      -- ^ A variable that caches knowledge about whether the merge has been
+      -- completed. If 'MergeCompleted', then we are sure the merge has been
+      -- completed, otherwise if 'MergeNotCompleted' we have to check the
+      -- 'MergingRunState'.
+      !(StrictMVar m (MergingRunState m h))
     -- | The result of merging a single run, is a single run.
   | SingleRun !(Run m h)
 
 newtype NumRuns = NumRuns { unNumRuns :: Int }
   deriving stock (Show, Eq)
 
+newtype UnspentCreditsVar s = UnspentCreditsVar { getUnspentCreditsVar :: PrimVar s Int }
+  deriving stock Eq
+
 data MergingRunState m h =
     CompletedMerge
       !(Run m h)
       -- ^ Output run
   | OngoingMerge
       !(V.Vector (Run m h))
       -- ^ Input runs
-      !(PrimVar (PrimState m) Int)
-      -- ^ The total number of performed merging steps.
-      !(PrimVar (PrimState m) Int)
-      -- ^ The total number of supplied credits.
+      !(SpentCreditsVar (PrimState m))
+      -- ^ The total number of spent credits.
       !(Merge m h)
 
+newtype TotalStepsVar s = TotalStepsVar { getTotalStepsVar ::  PrimVar s Int  }
+  deriving stock Eq
+
+newtype SpentCreditsVar s = SpentCreditsVar { getSpentCreditsVar :: PrimVar s Int }
+  deriving stock Eq
+
+data MergeCompleted = MergeCompleted | MergeNotCompleted
```

Comment:
If we only use this type for this one purpose, maybe we can be more explicit: `MergeKnownCompleted = MergeKnownCompleted | MergePotentiallyNotCompleted` or `MergeMaybeCompleted`.